sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachWriterTable.scala

```diff
@@ -141,17 +143,33 @@ class ForeachDataWriter[T](
       writer.process(rowConverter(record))
     } catch {
       case t: Throwable =>
-        writer.close(t)
+        closeWriter(t)
         throw t
     }
   }
 
   override def commit(): WriterCommitMessage = {
-    writer.close(null)
+    closeWriter(null)
     ForeachWriterCommitMessage
   }
 
-  override def abort(): Unit = {}
+  override def abort(): Unit = {
+    closeWriter(new RuntimeException("Foreach writer has been aborted"))
```

Comment:
I think it's better to change `DataWriter.abort` to pass the cause of failure in rather than creating this `RuntimeException`. An implementation may care about the error and do different stuff. @cloud-fan what do you think?

Reply:
@zsxwing I agree that it will be valuable to pass the original cause to the abort but i think that in order to due that, i will have to change the Utils.tryWithSafeFinallyAndFailureCallbacks such that the cause block will have the Throwable as an argument 
something like:
 def tryWithSafeFinallyAndFailureCallbacks[T](block: => T)
      (catchBlock: Throwable => Unit = _ => Unit, finallyBlock: => Unit = ()): T

it will require some further changes wherever the catch block is used

is it make sense?